### PR TITLE
Implement `list` context

### PR DESCRIPTION
**Pull request by hausdorff**
See commit messages for a detailed explanation of these features.

### CODE REVIEW

File: pkg/backend/backend.go [hunk 1]
```diff
@@ -222,3 +222,29 @@ func (c *backendClient) GetStackOutputs(ctx context.Context, name string) (resou
 	}
 	return res.Outputs, nil
 }
+
+// GetStackResourceOutputs returns the outputs of the stack with the given name.
+func (c *backendClient) GetStackResourceOutputs(ctx context.Context, name, typ string) (resource.PropertyMap, error) {
```
Comment (pgavlin) [inline pkg/backend/backend.go]: s/`GetStackResourceOutputs`/`GetResourceOutputs`/

File: pkg/backend/backend.go [hunk 2]
```diff
@@ -222,3 +222,29 @@ func (c *backendClient) GetStackOutputs(ctx context.Context, name string) (resou
...
+		return nil, err
+	}
+	if s == nil {
+		return nil, errors.Errorf("unknown stack \"%s\"", name)
+	}
+	snap, err := s.Snapshot(ctx)
+	if err != nil {
+		return nil, err
+	}
+	pm := resource.PropertyMap{}
```
Comment (pgavlin) [inline pkg/backend/backend.go]: Can we pull the common code between `GetStackOutputs` and this method out into a helper?

File: pkg/resource/deploy/builtins.go [hunk 3]
```diff
@@ -141,8 +141,16 @@ func (p *builtinProvider) Read(urn resource.URN, id resource.ID,
 
 func (p *builtinProvider) Invoke(tok tokens.ModuleMember,
 	args resource.PropertyMap) (resource.PropertyMap, []plugin.CheckFailure, error) {
+	if tok != "pulumi:pulumi:readStackResourceOutputs" {
```
Comment (pgavlin) [inline pkg/resource/deploy/builtins.go]: s/`readStackResourceOutputs`/`getResourceOutputs`

File: sdk/nodejs/query/index.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+export interface ListTypeArgs {
+    type: string;
+}
+
+export interface ListArgs extends ListTypeArgs {
+    stack?: string;
+}
+
+export interface ListContext {
+    // TODO: replace Observable with synchronous query model.
```
Comment (pgavlin) [inline sdk/nodejs/query/index.ts]: What's the idea here? This is not possible without some pretty nasty hacks if we build upon `invoke`.

File: pkg/backend/httpstate/backend.go [hunk 5]
```diff
@@ -1238,3 +1238,29 @@ func (c httpstateBackendClient) GetStackOutputs(ctx context.Context, name string
 
 	return backend.NewBackendClient(c.backend).GetStackOutputs(ctx, name)
 }
+
+// GetStackResourceOutputs returns the outputs of the stack with the given name.
+func (c httpstateBackendClient) GetStackResourceOutputs(ctx context.Context, name, typ string) (resource.PropertyMap, error) {
```
Comment (pgavlin) [inline pkg/backend/httpstate/backend.go]: Same naming + refactoring feedback here as above

File: pkg/resource/deploy/builtins.go [hunk 6]
```diff
@@ -174,3 +182,29 @@ func (p *builtinProvider) readStackReference(inputs resource.PropertyMap) (resou
 		"outputs": resource.NewObjectProperty(outputs),
 	}, nil
 }
+
+func (p *builtinProvider) readStackResourceOutputs(inputs resource.PropertyMap) (resource.PropertyMap, error) {
+	name, ok := inputs["stackName"]
+	contract.Assert(ok)
+	contract.Assert(name.IsString())
```
Comment (pgavlin) [inline pkg/resource/deploy/builtins.go]: These asserts really ought to turn into check failures

File: pkg/resource/deploy/builtins.go [hunk 7]
```diff
@@ -174,3 +182,29 @@ func (p *builtinProvider) readStackReference(inputs resource.PropertyMap) (resou
...
+
+	var typ string
+	if t, hasType := inputs["type"]; hasType {
+		contract.Assert(t.IsString())
```
Comment (pgavlin) [inline pkg/resource/deploy/builtins.go]: Same

**Comment by hausdorff** [comment]
Sorry for the churn on these PRs -- I'm going to close this and start pushing to the feature branch, and then open a PR from that feature branch for comments.

**Pull request closed without merging**